File: backendj/src/com/runsim/backend/SCTPTestServer.java
package com.runsim.backend;

import com.sun.nio.sctp.*;

import java.io.PrintStream;
import java.net.InetSocketAddress;
import java.nio.ByteBuffer;

public class SCTPTestServer {

    public void start(int port) throws Exception {
        InetSocketAddress serverAddress = new InetSocketAddress(port);
        SctpServerChannel ssc = SctpServerChannel.open();
        ssc.bind(serverAddress);
        System.out.println("server started listening on port " + port);

        ByteBuffer incomingBuffer = ByteBuffer.allocateDirect(1073741824);
        while (true) {
            SctpChannel sc = ssc.accept();

            AssociationHandler assocHandler = new AssociationHandler();
            MessageInfo inMessageInfo;
            while (true) {
                inMessageInfo = sc.receive(incomingBuffer, System.out, assocHandler);
                if (inMessageInfo == null || inMessageInfo.bytes() == -1) {
                    break;
                }

                byte[] receivedBytes = new byte[inMessageInfo.bytes()];
                for (int i = 0; i < receivedBytes.length; i++) {
                    receivedBytes[i] = incomingBuffer.get(i);
                }

                handleMessage(sc, receivedBytes);
            }
            sc.close();
        }
    }

    private void handleMessage(SctpChannel channel, byte[] data) throws Exception {
        MessageInfo msg = MessageInfo.createOutgoing(null, 0);
        msg.payloadProtocolID(60);
        channel.send(ByteBuffer.wrap(new byte[] { 1, 2, 3, 4 }), msg);
    }

    private static class AssociationHandler extends AbstractNotificationHandler<PrintStream> {

        @Override
        public HandlerResult handleNotification(AssociationChangeNotification notification, PrintStream attachment) {
            if (notification.event() == AssociationChangeNotification.AssocChangeEvent.COMM_UP) {
                int outbound = notification.association().maxOutboundStreams();
                int inbound = notification.association().maxInboundStreams();
                attachment.printf(
                        "New association setup with %d outbound streams" + ", and %d inbound streams.\n",
                        outbound, inbound
                );
            }
            return HandlerResult.CONTINUE;
        }

        @Override
        public HandlerResult handleNotification(ShutdownNotification notification, PrintStream attachment) {
            attachment.print("The association has been shutdown.\n");
            return HandlerResult.RETURN;
        }
    }
}

File: backend/src/main/kotlin/SCTPTestServer.kt
import com.sun.nio.sctp.*

import java.*
import java.io.*
import java.net.*
import java.nio.*
import java.nio.charset.*
import java.text.*
import java.util.*

object SCTPTestServer {

    @JvmStatic
    fun main(args: Array<String>) {
        val serverAddr = InetSocketAddress(3457)
        val ssc = SctpServerChannel.open()
        ssc.bind(serverAddr)

        val recvbuf = ByteBuffer.allocateDirect(255)
        while (true) {
            val sc = ssc.accept()

            val outMsgInfo = MessageInfo.createOutgoing(null, 0)
            outMsgInfo.payloadProtocolID(60)
            sc.send(ByteBuffer.wrap(byteArrayOf(45, 45, 45, 45)), outMsgInfo)
            // shutdown and receive all pending messages/notifications
            sc.shutdown()
            val assocHandler = AssociationHandler()
            var inMessageInfo: MessageInfo? = null
            while (true) {
                inMessageInfo = sc.receive(recvbuf, System.out, assocHandler)
                if (inMessageInfo == null || inMessageInfo.bytes() == -1) {
                    break
                }
            }
            sc.close()
        }
    }

    private class AssociationHandler : AbstractNotificationHandler<PrintStream>() {
        override fun handleNotification(not: AssociationChangeNotification, stream: PrintStream?): HandlerResult {
            if (not.event() == AssociationChangeNotification.AssocChangeEvent.COMM_UP) {
                val outbound = not.association().maxOutboundStreams()
                val inbound = not.association().maxInboundStreams()
                stream!!.printf(
                    "New association setup with %d outbound streams" + ", and %d inbound streams.\n",
                    outbound, inbound
                )
            }
            return HandlerResult.CONTINUE
        }
        override fun handleNotification(not: ShutdownNotification?, stream: PrintStream): HandlerResult {
            stream.printf("The association has been shutdown.\n")
            return HandlerResult.RETURN
        }
    }
}
